tensorflow_probability/python/math/special.py

```diff
@@ -521,6 +545,332 @@ def betainc(a, b, x, name=None):
+  """Computes an initial approximation for `betaincinv(a, b, y)`."""
+  numpy_dtype = dtype_util.as_numpy_dtype(dtype)
+  tiny = tf.constant(np.finfo(numpy_dtype).tiny, dtype=dtype)
+  eps = tf.constant(np.finfo(numpy_dtype).eps, dtype=dtype)
+  one = tf.constant(1., dtype=dtype)
+  two = tf.constant(2., dtype=dtype)
+  three = tf.constant(3., dtype=dtype)
+  five = tf.constant(5., dtype=dtype)
+  six = tf.constant(6., dtype=dtype)
+  min_log = tf.math.log(
+      tf.constant(2. ** np.finfo(numpy_dtype).minexp, dtype))
+  max_log = tf.math.log(
+      tf.constant(2. ** (np.finfo(numpy_dtype).maxexp - 1.), dtype))
+
+  # When min(a, b) >= 1, we use the approximation proposed by [1].
+
+  # Equation 26.5.22 [1, page 945].
+  yp = -tf.math.ndtri(y)
+  inv_2a_minus_one = tf.math.reciprocal(two * a - one)
+  inv_2b_minus_one = tf.math.reciprocal(two * b - one)
+  lmb = (tf.math.square(yp) - three) / six
+  h = two * tf.math.reciprocal(inv_2a_minus_one + inv_2b_minus_one)
+  w = (yp * tf.math.sqrt(h + lmb) / h -
+      (inv_2b_minus_one - inv_2a_minus_one) *
+      (lmb + five / six - two / (three * h)))
+  result_for_large_a_and_b = a / (a + b * tf.math.exp(two * w))
+
+  # When min(a, b) < 1 and max(a, b) >= 1, we use the approximation proposed by
+  # [2]. This approximation depends on the following approximation for betainc:
+  #   betainc(a, b, x) ~=
+  #       x ** a / (integral_approx * a) , when x <= mean ,
+  #       (1 - x) ** b / (integral_approx * b) , when x > mean ,
+  # where:
+  #   integral_approx = (mean ** a) / a + (mean_complement ** b) / b ,
+  #   mean = a / (a + b) ,
+  #   mean_complement = 1 - mean = b / (a + b) .
+  # We invert betainc(a, b, x) with respect to x in the proper regime.
+
+  # Equation 6.4.7 [2, page 271].
+  a_plus_b = a + b
+  mean = a / a_plus_b
+  mean_complement = b / a_plus_b
+  integral_approx_part_a = tf.math.exp(tf.math.xlogy(a, mean) - tf.math.log(a))
+  integral_approx_part_b = tf.math.exp(tf.math.xlogy(b, mean_complement) -
+      tf.math.log(b))
+  integral_approx = integral_approx_part_a + integral_approx_part_b
+
+  # Solve Equation 6.4.8 [2, page 271] for x in the respective regimes.
+  inv_a = tf.math.reciprocal(a)
+  inv_b = tf.math.reciprocal(b)
+  result_for_small_a_or_b = tf.where(
+      tf.math.less_equal(y, (integral_approx_part_a / integral_approx)),
+      tf.math.exp(tf.math.xlogy(inv_a, y) + tf.math.xlogy(inv_a, a) +
+          tf.math.xlogy(inv_a, integral_approx)),
+      one - tf.math.exp(tf.math.xlog1py(inv_b, -y) + tf.math.xlogy(inv_b, b) +
+          tf.math.xlogy(inv_b, integral_approx)))
+
+  # And when max(a, b) < 1, we use the approximation proposed by [3] for the
+  # same domain:
+  #   betaincinv(a, b, y) ~= xg / (1 + xg) ,
+  # where:
+  #   xg = (a * y * Beta(a, b)) ** (1 / a) .
+  log_xg = tf.math.xlogy(inv_a, a) + tf.math.xlogy(inv_a, y) + (
+      inv_a * lbeta(a, b))
+  xg = tf.math.exp(tf.clip_by_value(log_xg, min_log, max_log))
```

Comment:
There seems to be quite a bit of clipping in this code (here and a few places below). Are these necessary / can they be removed in favor of `tf.where` guarding code earlier up?

Reply:
There were four uses of `tf.clip_by_value` in the code (I replaced two of them). They handled cases that depend on the combination of values for the parameters `a`, `b`, and `y`. I don't know if it's possible (and convenient) to list and describe all these problematic combinations of values and then use `tf.where` to guard the code from the respective cases.

The first use of `tf.clip_by_value` (line 629) avoided overflow in `tf.math.exp(log_xg)`: for example, when `a` and `b` are very small and `x` is close enough to `1`. But in this case, I replaced `tf.clip_by_value` with `tf.math.minimum`.

The second use of `tf.clip_by_value` (line 641) ensures that the result of `_betaincinv_initial_approx` is always greater than `0` and less than `1` and, consequently, that the condition `initial_low < initial_candidate < initial_high` is always satisfied. For example, if

```
a, b, y, dtype = (3.5565727, 0.011408293, 0.130523313, float32)
```

then `initial_candidate == 1.` and `initial_candidate == initial_high`.

The reason for the third use of `tf.clip_by_value` (line 714) is commented out in the code.

The last use of `tf.clip_by_value` (line 764) prevented `one - result` from being rounded up to `1` when `use_symmetry_relation` and `result < eps`. But in this case, I replaced `tf.clip_by_value` with `tf.math.maximum`.

Reply:
Thanks for the explanation, those uses seem pretty reasonable to me.